heirloom/heirloom-pkgtools/man/cktime.1
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"  Copyright 1989 AT&T  Copyright (c) 1992, Sun Microsystems, Inc.  All Rights Reserved
.\" Portions Copyright (c) 2007 Gunnar Ritter, Freiburg i. Br., Germany
.\"
.\" Sccsid @(#)cktime.1	1.5 (gritter) 3/4/07
.\"
.\" from OpenSolaris cktime 1 "14 Sep 1992" "SunOS 5.11" "User Commands"
.TH CKTIME 1 "3/4/07" "Heirloom Packaging Tools" "User Commands"
.SH NAME
cktime, errtime, helptime, valtime \- display a prompt; verify and return a time of day
.SH SYNOPSIS
.HP
.ad l
.nh
\fBcktime\fR [\fB\-Q\fR] [\fB\-W\fR \fIwidth\fR] [\fB\-f\fR \fIformat\fR] [\fB\-d\fR \fIdefault\fR] [\fB\-h\fR \fIhelp\fR]
[\fB\-e\fR \fIerror\fR] [\fB\-p\fR \fIprompt\fR] [\fB\-k\fR \fIpid\fR [\fB\-s\fR \fIsignal\fR]]
.HP
.PD 0
.ad l
\fBerrtime\fR [\fB\-W\fR \fIwidth\fR] [\fB\-e\fR \fIerror\fR] [\fB\-f\fR \fIformat\fR]
.HP
.PD 0
.ad l
\fBhelptime\fR [\fB\-W\fR \fIwidth\fR] [\fB\-h\fR \fIhelp\fR] [\fB\-f\fR \fIformat\fR]
.HP
.PD 0
.ad l
\fBvaltime\fR [\fB\-f\fR \fIformat\fR] \fIinput\fR
.br
.PD
.ad b
.hy 1
.SH DESCRIPTION
The \fBcktime\fR utility prompts a user and validates the response.
It defines, among other things, a prompt message whose response should be a time, text for help and error messages,
and a default value (which is returned if the user responds with a RETURN).
The user response must match the defined format for the time of day.
.PP
All messages are limited in length to 70 characters and are formatted automatically.
Any white space used in the definition (including \fBNEWLINE)\fR is stripped.
The \fB\-W\fR option cancels the automatic formatting.
When a tilde is placed at the beginning or end of a message definition, the default text is inserted at that point, allowing both custom text and the
default text to be displayed.
.PP
If the prompt, help or error message is not defined, the default message (as defined under \fBNOTES\fR) is displayed.
.PP
Three visual tool modules are linked to the \fBcktime\fR command.
They are \fBerrtime\fR (which formats and displays an error message), \fBhelptime\fR (which
formats and displays a help message), and \fBvaltime\fR (which validates a response).
These modules should be used in conjunction with \fBFML\fR objects.
In this
instance, the \fBFML\fR object defines the prompt.
When \fBformat\fR is defined in the \fBerrtime\fR and \fBhelptime\fR modules, the messages
will describe the expected format.
.PP
The following options and operands are supported:
.TP
\fB\-d\fR\fI default\fR
Defines the default value as \fIdefault\fR.
The default is not validated and so does not have to meet any criteria.
.TP
\fB\-e\fR \fIerror\fR
Defines the error message as \fIerror\fR.
.TP
\fB\-f\fR \fIformat\fR
Specifies the format against which the input will be verified.
Possible formats and their definitions are:
.RS
.nf
\fB%H\fR  =  hour (00 - 23)
\fB%I\fR  =  hour (00 - 12)
\fB%M\fR  =  minute (00 - 59)
\fB%p\fR  =  ante meridian or post meridian
\fB%r\fR  =  time as \fB%I:%M:%S %p\fR
\fB%R\fR  =  time as \fB%H:%M\fR (the default format)
\fB%S\fR  =  seconds (00 - 59)
\fB%T\fR  =  time as \fB%H:%M:%S\fR
.fi
.RE
.TP
\fB\-h\fR \fIhelp\fR
Defines the help messages as \fIhelp\fR.
.TP
\fB\-k\fR\fI pid\fR
Specifies that process \fBID\fR \fIpid\fR is to be sent a signal
if the user chooses to abort.
.TP
\fB\-p\fR\fI prompt\fR
Defines the prompt message as \fIprompt\fR.
.TP
\fB\-Q\fR
Specifies that quit will not be allowed as a valid response.
.TP
\fB\-s\fR \fIsignal\fR
Specifies that the process \fBID\fR \fIpid\fR defined with
the \fB\-k\fR option is to be sent signal \fIsignal\fR when quit is chosen.
If no signal is specified, SIGTERM is used.
.TP
\fB\-W\fR\fI width\fR
Specifies that prompt, help and error messages will be formatted to a line length of \fIwidth\fR.
.TP
\fB\fIinput\fR
Input to be verified against format criteria.
.SH EXIT STATUS
The following exit values are returned:
.PD 0
.TP
.B 0
Successful execution.
.TP
.B 1
\fBEOF\fR on input, or negative width on \fB\-W\fR option, or usage error.
.TP
.B 3
User termination (quit).
.TP
.B 4
Garbled format argument.
.PD
.SH NOTES
The default prompt for \fBcktime\fR is:
.PP
.in +2
.nf
Enter a time of day [?,q]:
.fi
.in -2
.PP
The default error message is:
.PP
.in +2
.nf
ERROR: Please enter the time of day.
Format is <format>.
.fi
.in -2
.PP
The default help message is:
.PP
.in +2
.nf
Please enter the time of day.
Format is <format>.
.fi
.in -2
.PP
When the quit option is chosen (and allowed), \fBq\fR is returned along with the return code \fB3\fR.
The \fBvaltime\fR module will not produce any output.
It returns \fB0\fR for success and non-zero for failure.


heirloom/heirloom-pkgtools/valtools/cktime.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*	from OpenSolaris "cktime.c	1.10	05/06/08 SMI"	*/

/*
 * Portions Copyright (c) 2007 Gunnar Ritter, Freiburg i. Br., Germany
 *
 * Sccsid @(#)cktime.c	1.3 (gritter) 3/3/07
 */

#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <stdlib.h>
#include <locale.h>
#include <libintl.h>
#include <limits.h>
#include <unistd.h>
#include "usage.h"
#include "libadm.h"

#define	BADPID	(-2)

static char	*prog;
static char	*deflt = NULL, *prompt = NULL, *error = NULL, *help = NULL;
static int	kpid = BADPID;
static int	signo;
static char	*fmt;

static const char	vusage[] = "f";
static const char	husage[] = "fWh";
static const char	eusage[] = "fWe";

#define	MYFMT \
	"%s:ERROR:invalid format\n" \
	"valid format descriptors are:\n" \
	"\t%%H  #hour (00-23)\n" \
	"\t%%I  #hour (00-12)\n" \
	"\t%%M  #minute (00-59)\n" \
	"\t%%p  #AM, PM, am or pm\n" \
	"\t%%r  #time as %%I:%%M:%%S %%p\n" \
	"\t%%R  #time as %%H:%%M (default)\n" \
	"\t%%S  #seconds (00-59)\n" \
	"\t%%T  #time as %%H:%%M:%%S\n"

static void
usage(void)
{
	switch (*prog) {
	default:
		(void) fprintf(stderr,
			gettext("usage: %s [options] [-f format]\n"),
			prog);
		(void) fprintf(stderr, gettext(OPTMESG));
		(void) fprintf(stderr, gettext(STDOPTS));
		break;

	case 'v':
		(void) fprintf(stderr,
			gettext("usage: %s [-f format] input\n"), prog);
		break;

	case 'h':
		(void) fprintf(stderr,
			gettext("usage: %s [options] [-f format]\n"),
			prog);
		(void) fprintf(stderr, gettext(OPTMESG));
		(void) fprintf(stderr,
			gettext("\t-W width\n\t-h help\n"));
		break;

	case 'e':
		(void) fprintf(stderr,
			gettext("usage: %s [options] [-f format]\n"),
			prog);
		(void) fprintf(stderr, gettext(OPTMESG));
		(void) fprintf(stderr,
			gettext("\t-W width\n\t-e error\n"));
		break;
	}
	exit(1);
}

/*
 * Given argv[0], return a pointer to the basename of the program.
 */
static char *
prog_name(char *arg0)
{
	char *str;

	/* first strip trailing '/' characters (exec() allows these!) */
	str = arg0 + strlen(arg0);
	while (str > arg0 && *--str == '/')
		*str = '\0';
	if ((str = strrchr(arg0, '/')) != NULL)
		return (str + 1);
	return (arg0);
}

int
main(int argc, char **argv)
{
	int c, n;
	char *tod;
	size_t	len;

	(void) setlocale(LC_ALL, "");

#if	!defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN	"SYS_TEST"
#endif
	(void) textdomain(TEXT_DOMAIN);

	prog = prog_name(argv[0]);

	while ((c = getopt(argc, argv, "f:d:p:e:h:k:s:QW:?")) != EOF) {
		/* check for invalid option */
		if ((*prog == 'v') && !strchr(vusage, c))
			usage();
		if ((*prog == 'e') && !strchr(eusage, c))
			usage();
		if ((*prog == 'h') && !strchr(husage, c))
			usage();

		switch (c) {
		case 'Q':
			ckquit = 0;
			break;

		case 'W':
			ckwidth = atoi(optarg);
			if (ckwidth < 0) {
				(void) fprintf(stderr,
		gettext("%s: ERROR: negative display width specified\n"),
					prog);
				exit(1);
			}
			break;

		case 'f':
			fmt = optarg;
			break;

		case 'd':
			deflt = optarg;
			break;

		case 'p':
			prompt = optarg;
			break;

		case 'e':
			error = optarg;
			break;

		case 'h':
			help = optarg;
			break;

		case 'k':
			kpid = atoi(optarg);
			break;

		case 's':
			signo = atoi(optarg);
			break;

		default:
			usage();
		}
	}

	if (signo) {
		if (kpid == BADPID)
			usage();
	} else
		signo = SIGTERM;

	if (*prog == 'v') {
		if (argc != (optind+1))
			usage();
		n = cktime_val(fmt, argv[optind]);
		/*
		 * TRANSLATION_NOTE
		 * In the below, "AM", "PM", "am", and "pm" are
		 * keywords.  So, do not translate them.
		 */
		if (n == 4)
			(void) fprintf(stderr, gettext(MYFMT), prog);
		exit(n);
	}

	if (optind != argc)
		usage();

	if (*prog == 'e') {
		ckindent = 0;
		if (cktime_err(fmt, error)) {
			(void) fprintf(stderr, gettext(MYFMT), prog);
			exit(4);
		} else
			exit(0);
	} else if (*prog == 'h') {
		ckindent = 0;
		if (cktime_hlp(fmt, help)) {
			(void) fprintf(stderr, gettext(MYFMT), prog);
			exit(4);
		} else
			exit(0);
	}

	if (deflt) {
		len = strlen(deflt) + 1;
		if (len < MAX_INPUT)
			len = MAX_INPUT;
	} else {
		len = MAX_INPUT;
	}
	tod = (char *)malloc(len);
	if (!tod) {
		(void) fprintf(stderr,
			gettext("Not enough memory\n"));
		exit(1);
	}
	n = cktime(tod, fmt, deflt, error, help, prompt);
	if (n == 3) {
		if (kpid > -2)
			(void) kill(kpid, signo);
		(void) puts("q");
	} else if (n == 0)
		(void) fputs(tod, stdout);
	if (n == 4)
		(void) fprintf(stderr, gettext(MYFMT), prog);
	return (n);
}
